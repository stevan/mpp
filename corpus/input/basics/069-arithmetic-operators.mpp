# Subtraction
10 - 5;
$x - $y;
100 - 50 - 25;
5.5 - 2.3;
0 - 10;  # negation

# Multiplication
5 * 6;
$x * $y;
2 * 3 * 4;
3.14 * 2;
-5 * -5;

# Division
10 / 2;
$x / $y;
100 / 4 / 5;
7.5 / 2.5;
1 / 3;  # fractional result

# Modulo
10 % 3;
$x % $y;
17 % 5;
100 % 7;
-10 % 3;

# Mixed operations
10 + 5 - 3;
2 * 3 + 4;
10 / 2 * 5;
100 - 10 * 5;

# Complex expressions
(10 + 5) * 2;
100 / (20 - 10);
($x + $y) * ($a - $b);
((10 + 5) * 3) - (20 / 4);

# Assignment with operators
my $result = 10 + 20 * 3;
my $average = ($a + $b + $c) / 3;
my $remainder = $total % $divisor;

# Numeric assignment operators
$x += 10;
$y -= 5;
$z *= 2;
$w /= 4;
$m %= 3;