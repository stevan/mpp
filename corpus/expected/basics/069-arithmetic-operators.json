[
  {
    "type": "BinaryOp",
    "operator": "-",
    "left": {
      "type": "Number",
      "value": "10"
    },
    "right": {
      "type": "Number",
      "value": "5"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "-",
    "left": {
      "type": "Variable",
      "name": "$x"
    },
    "right": {
      "type": "Variable",
      "name": "$y"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "-",
    "left": {
      "type": "BinaryOp",
      "operator": "-",
      "left": {
        "type": "Number",
        "value": "100"
      },
      "right": {
        "type": "Number",
        "value": "50"
      }
    },
    "right": {
      "type": "Number",
      "value": "25"
    }
  },
  {
    "type": "BinaryOp",
    "operator": ".",
    "left": {
      "type": "BinaryOp",
      "operator": "-",
      "left": {
        "type": "BinaryOp",
        "operator": ".",
        "left": {
          "type": "Number",
          "value": "5"
        },
        "right": {
          "type": "Number",
          "value": "5"
        }
      },
      "right": {
        "type": "Number",
        "value": "2"
      }
    },
    "right": {
      "type": "Number",
      "value": "3"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "-",
    "left": {
      "type": "Number",
      "value": "0"
    },
    "right": {
      "type": "Number",
      "value": "10"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "*",
    "left": {
      "type": "Number",
      "value": "5"
    },
    "right": {
      "type": "Number",
      "value": "6"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "*",
    "left": {
      "type": "Variable",
      "name": "$x"
    },
    "right": {
      "type": "Variable",
      "name": "$y"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "*",
    "left": {
      "type": "BinaryOp",
      "operator": "*",
      "left": {
        "type": "Number",
        "value": "2"
      },
      "right": {
        "type": "Number",
        "value": "3"
      }
    },
    "right": {
      "type": "Number",
      "value": "4"
    }
  },
  {
    "type": "BinaryOp",
    "operator": ".",
    "left": {
      "type": "Number",
      "value": "3"
    },
    "right": {
      "type": "BinaryOp",
      "operator": "*",
      "left": {
        "type": "Number",
        "value": "14"
      },
      "right": {
        "type": "Number",
        "value": "2"
      }
    }
  },
  {
    "type": "BinaryOp",
    "operator": "*",
    "left": {
      "type": "UnaryOp",
      "operator": "-",
      "operand": {
        "type": "Number",
        "value": "5"
      }
    },
    "right": {
      "type": "UnaryOp",
      "operator": "-",
      "operand": {
        "type": "Number",
        "value": "5"
      }
    }
  },
  {
    "type": "BinaryOp",
    "operator": "/",
    "left": {
      "type": "Number",
      "value": "10"
    },
    "right": {
      "type": "Number",
      "value": "2"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "/",
    "left": {
      "type": "Variable",
      "name": "$x"
    },
    "right": {
      "type": "Variable",
      "name": "$y"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "/",
    "left": {
      "type": "BinaryOp",
      "operator": "/",
      "left": {
        "type": "Number",
        "value": "100"
      },
      "right": {
        "type": "Number",
        "value": "4"
      }
    },
    "right": {
      "type": "Number",
      "value": "5"
    }
  },
  {
    "type": "BinaryOp",
    "operator": ".",
    "left": {
      "type": "BinaryOp",
      "operator": ".",
      "left": {
        "type": "Number",
        "value": "7"
      },
      "right": {
        "type": "BinaryOp",
        "operator": "/",
        "left": {
          "type": "Number",
          "value": "5"
        },
        "right": {
          "type": "Number",
          "value": "2"
        }
      }
    },
    "right": {
      "type": "Number",
      "value": "5"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "/",
    "left": {
      "type": "Number",
      "value": "1"
    },
    "right": {
      "type": "Number",
      "value": "3"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "%",
    "left": {
      "type": "Number",
      "value": "10"
    },
    "right": {
      "type": "Number",
      "value": "3"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "%",
    "left": {
      "type": "Variable",
      "name": "$x"
    },
    "right": {
      "type": "Variable",
      "name": "$y"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "%",
    "left": {
      "type": "Number",
      "value": "17"
    },
    "right": {
      "type": "Number",
      "value": "5"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "%",
    "left": {
      "type": "Number",
      "value": "100"
    },
    "right": {
      "type": "Number",
      "value": "7"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "%",
    "left": {
      "type": "UnaryOp",
      "operator": "-",
      "operand": {
        "type": "Number",
        "value": "10"
      }
    },
    "right": {
      "type": "Number",
      "value": "3"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "-",
    "left": {
      "type": "BinaryOp",
      "operator": "+",
      "left": {
        "type": "Number",
        "value": "10"
      },
      "right": {
        "type": "Number",
        "value": "5"
      }
    },
    "right": {
      "type": "Number",
      "value": "3"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "+",
    "left": {
      "type": "BinaryOp",
      "operator": "*",
      "left": {
        "type": "Number",
        "value": "2"
      },
      "right": {
        "type": "Number",
        "value": "3"
      }
    },
    "right": {
      "type": "Number",
      "value": "4"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "*",
    "left": {
      "type": "BinaryOp",
      "operator": "/",
      "left": {
        "type": "Number",
        "value": "10"
      },
      "right": {
        "type": "Number",
        "value": "2"
      }
    },
    "right": {
      "type": "Number",
      "value": "5"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "-",
    "left": {
      "type": "Number",
      "value": "100"
    },
    "right": {
      "type": "BinaryOp",
      "operator": "*",
      "left": {
        "type": "Number",
        "value": "10"
      },
      "right": {
        "type": "Number",
        "value": "5"
      }
    }
  },
  {
    "type": "BinaryOp",
    "operator": "*",
    "left": {
      "type": "BinaryOp",
      "operator": "+",
      "left": {
        "type": "Number",
        "value": "10"
      },
      "right": {
        "type": "Number",
        "value": "5"
      }
    },
    "right": {
      "type": "Number",
      "value": "2"
    }
  },
  {
    "type": "BinaryOp",
    "operator": "/",
    "left": {
      "type": "Number",
      "value": "100"
    },
    "right": {
      "type": "BinaryOp",
      "operator": "-",
      "left": {
        "type": "Number",
        "value": "20"
      },
      "right": {
        "type": "Number",
        "value": "10"
      }
    }
  },
  {
    "type": "BinaryOp",
    "operator": "*",
    "left": {
      "type": "BinaryOp",
      "operator": "+",
      "left": {
        "type": "Variable",
        "name": "$x"
      },
      "right": {
        "type": "Variable",
        "name": "$y"
      }
    },
    "right": {
      "type": "BinaryOp",
      "operator": "-",
      "left": {
        "type": "Variable",
        "name": "$a"
      },
      "right": {
        "type": "Variable",
        "name": "$b"
      }
    }
  },
  {
    "type": "BinaryOp",
    "operator": "-",
    "left": {
      "type": "BinaryOp",
      "operator": "*",
      "left": {
        "type": "BinaryOp",
        "operator": "+",
        "left": {
          "type": "Number",
          "value": "10"
        },
        "right": {
          "type": "Number",
          "value": "5"
        }
      },
      "right": {
        "type": "Number",
        "value": "3"
      }
    },
    "right": {
      "type": "BinaryOp",
      "operator": "/",
      "left": {
        "type": "Number",
        "value": "20"
      },
      "right": {
        "type": "Number",
        "value": "4"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$result"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "+",
      "left": {
        "type": "Number",
        "value": "10"
      },
      "right": {
        "type": "BinaryOp",
        "operator": "*",
        "left": {
          "type": "Number",
          "value": "20"
        },
        "right": {
          "type": "Number",
          "value": "3"
        }
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$average"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "/",
      "left": {
        "type": "BinaryOp",
        "operator": "+",
        "left": {
          "type": "BinaryOp",
          "operator": "+",
          "left": {
            "type": "Variable",
            "name": "$a"
          },
          "right": {
            "type": "Variable",
            "name": "$b"
          }
        },
        "right": {
          "type": "Variable",
          "name": "$c"
        }
      },
      "right": {
        "type": "Number",
        "value": "3"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$remainder"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "%",
      "left": {
        "type": "Variable",
        "name": "$total"
      },
      "right": {
        "type": "Variable",
        "name": "$divisor"
      }
    }
  },
  {
    "type": "Assignment",
    "left": {
      "type": "Variable",
      "name": "$x"
    },
    "operator": "+=",
    "right": {
      "type": "Number",
      "value": "10"
    }
  },
  {
    "type": "Assignment",
    "left": {
      "type": "Variable",
      "name": "$y"
    },
    "operator": "-=",
    "right": {
      "type": "Number",
      "value": "5"
    }
  },
  {
    "type": "Assignment",
    "left": {
      "type": "Variable",
      "name": "$z"
    },
    "operator": "*=",
    "right": {
      "type": "Number",
      "value": "2"
    }
  },
  {
    "type": "Assignment",
    "left": {
      "type": "Variable",
      "name": "$w"
    },
    "operator": "/=",
    "right": {
      "type": "Number",
      "value": "4"
    }
  },
  {
    "type": "Assignment",
    "left": {
      "type": "Variable",
      "name": "$m"
    },
    "operator": "%=",
    "right": {
      "type": "Number",
      "value": "3"
    }
  }
]